src/dshould/prettyprint.d

```diff
@@ -0,0 +1,385 @@
+module dshould.prettyprint;
+
+import std.algorithm;
+import std.range;
+import std.typecons;
+
+/**
+ * This function takes the input text and returns a pretty-printed, multiline, indented version.
+ * It assumes that the input text is the output of toString and forms a valid comma separated paren tree.
+ *
+ * A comma separated paren tree is a string that contains a balanced number of quotation marks, parentheses
+ * and brackets.
+ */
+public string prettyprint(const string text, size_t columnLength = 80)
+{
+    const tree = text.parse;
+
+    if (tree.isNull)
+    {
+        return text;
+    }
+
+    return tree.get.prettyprint(columnLength);
+}
+
+///
+@("pretty print a string")
+unittest
+{
+    import dshould : equal, should;
+
+    prettyprint("Foo()").should.equal("Foo()");
+    prettyprint("Foo(Bar(Baz()), Baq())", 16).should.equal(
+"Foo(
+    Bar(Baz()),
+    Baq()
+)");
+    prettyprint("Foo(Bar(Baz()), Baq())", 13).should.equal(
+"Foo(
+    Bar(
+        Baz()
+    ),
+    Baq()
+)");
+}
+
+private enum indent = "    ";
+
+private string prettyprint(const Tree tree, size_t width)
+{
+    import std.string : stripLeft;
+
+    Appender!string result;
+
+    // skip prefix so caller can decide whether or not to strip
+    void walkOneLine(const Tree tree)
+    {
+        if (tree.parenType.isNull)
+        {
+            return;
+        }
+        result ~= tree.parenType.get.opening;
+        tree.children.enumerate.each!((index, child) {
+            if (index > 0)
+            {
+                result ~= ",";
+            }
+            result ~= child.prefix;
+            walkOneLine(child);
+        });
+        result ~= tree.parenType.get.closing;
+    }
+
+    void walk(size_t level, const Tree tree)
+    {
+        if (!tree.lengthExceeds(width - level * indent.length))
+        {
+            result ~= tree.prefix.stripLeft;
+            walkOneLine(tree);
+            return;
+        }
+
+        result ~= tree.prefix.stripLeft;
+        if (tree.parenType.isNull)
+        {
+            return;
+        }
+        result ~= tree.parenType.get.opening;
+        tree.children.enumerate.each!((index, child) {
+            if (index > 0)
+            {
+                result ~= ",";
+            }
+            result ~= "\n";
+            (level + 1).iota.each!((_) { result ~= indent; });
```

Comment:
Why not `repeat`? Comment?

Reply:
`repeat` and `join` force an extra allocation.